File: backend/operations-service/src/main/java/com/poscl/operations/application/service/ReservationsService.java
package com.poscl.operations.application.service;

import com.poscl.operations.domain.entity.Reservation;
import com.poscl.operations.domain.entity.RestaurantTable;
import com.poscl.operations.domain.repository.ReservationRepository;
import com.poscl.operations.domain.repository.RestaurantTableRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class ReservationsService {

    private final ReservationRepository reservationRepository;
    private final RestaurantTableRepository tableRepository;
    private final AutomationService automationService;
    private final CrmSyncService crmSyncService;

    // ==================== RESERVATIONS ====================

    public List<Reservation> getReservationsForDate(UUID tenantId, LocalDate date) {
        return reservationRepository.findByTenantIdAndFechaOrderByHoraAsc(tenantId, date);
    }

    public List<Reservation> getReservationsForDateAndBranch(UUID tenantId, UUID branchId, LocalDate date) {
        return reservationRepository.findByTenantIdAndBranchIdAndFechaOrderByHoraAsc(tenantId, branchId, date);
    }

    public Optional<Reservation> getReservationById(UUID id) {
        return reservationRepository.findById(id);
    }

    @Transactional
    public Reservation createReservation(UUID tenantId, UUID branchId, String clienteNombre,
            String clienteTelefono, LocalDate fecha, LocalTime hora,
            int personas, UUID tableId, String notas, String serviceType) {

        Reservation reservation = Reservation.builder()
                .tenantId(tenantId)
                .branchId(branchId)
                .clienteNombre(clienteNombre)
                .clienteTelefono(clienteTelefono)
                .fecha(fecha)
                .hora(hora)
                .personas(personas)
                .notas(notas)
                .serviceType(serviceType != null ? serviceType : "MESA")
                .build();

        if (tableId != null) {
            RestaurantTable table = tableRepository.findById(tableId).orElse(null);
            if (table != null) {
                reservation.setTable(table);
            }
        }

        log.info("Creating reservation for {} on {} at {}", clienteNombre, fecha, hora);
        Reservation saved = reservationRepository.save(reservation);

        // Trigger "Nueva Reserva" automation
        automationService.triggerAutomations(tenantId, "nueva-reserva", saved);

        // Sync customer with CRM
        crmSyncService.syncReservationCustomer(tenantId, clienteNombre, clienteTelefono, null, saved.getId());

        return saved;
    }

    @Transactional
    public Reservation updateStatus(UUID reservationId, String newStatus) {
        Reservation reservation = reservationRepository.findById(reservationId)
                .orElseThrow(() -> new IllegalArgumentException("Reservation not found: " + reservationId));

        String oldStatus = reservation.getEstado();

        switch (newStatus.toUpperCase()) {
            case "CONFIRMADA" -> reservation.confirm();
            case "EN_CURSO" -> reservation.start();
            case "COMPLETADA" -> reservation.complete();
            case "CANCELADA" -> reservation.cancel();
            case "NO_SHOW" -> reservation.markNoShow();
            default -> reservation.setEstado(newStatus);
        }

        log.info("Updated reservation {} status to {}", reservationId, newStatus);
        Reservation saved = reservationRepository.save(reservation);

        // Trigger status-based automations
        if (!oldStatus.equalsIgnoreCase(newStatus)) {
            if ("CONFIRMADA".equalsIgnoreCase(newStatus)) {
                automationService.triggerAutomations(reservation.getTenantId(), "confirmacion", saved);
            } else if ("CANCELADA".equalsIgnoreCase(newStatus)) {
                automationService.triggerAutomations(reservation.getTenantId(), "cancelacion", saved);
                crmSyncService.recordCancellation(reservation.getTenantId(), reservation.getClienteTelefono(),
                        saved.getId());
            } else if ("COMPLETADA".equalsIgnoreCase(newStatus)) {
                automationService.triggerAutomations(reservation.getTenantId(), "completada", saved);
                crmSyncService.recordCompletedVisit(reservation.getTenantId(), reservation.getClienteTelefono(),
                        reservation.getClienteNombre(), saved.getId());
            }
        }

        return saved;
    }

    @Transactional
    public Reservation assignTable(UUID reservationId, UUID tableId) {
        Reservation reservation = reservationRepository.findById(reservationId)
                .orElseThrow(() -> new IllegalArgumentException("Reservation not found: " + reservationId));

        RestaurantTable table = tableRepository.findById(tableId)
                .orElseThrow(() -> new IllegalArgumentException("Table not found: " + tableId));

        reservation.setTable(table);
        return reservationRepository.save(reservation);
    }

    // ==================== TABLES ====================

    public List<RestaurantTable> getTables(UUID tenantId, UUID branchId) {
        return tableRepository.findByTenantIdAndBranchIdAndActivoOrderByNumeroAsc(tenantId, branchId, true);
    }

    public List<RestaurantTable> getAvailableTables(UUID tenantId, UUID branchId, int minCapacity) {
        return tableRepository.findByTenantIdAndBranchIdAndCapacidadGreaterThanEqualAndEstado(
                tenantId, branchId, minCapacity, "DISPONIBLE");
    }

    @Transactional
    public RestaurantTable updateTableStatus(UUID tableId, String newStatus) {
        RestaurantTable table = tableRepository.findById(tableId)
                .orElseThrow(() -> new IllegalArgumentException("Table not found: " + tableId));

        switch (newStatus.toUpperCase()) {
            case "DISPONIBLE" -> table.release();
            case "OCUPADA" -> table.occupy();
            case "RESERVADA" -> table.reserve();
            default -> table.setEstado(newStatus);
        }

        log.info("Updated table {} status to {}", table.getNumero(), newStatus);
        return tableRepository.save(table);
    }

    // ==================== STATS ====================

    public ReservationStats getStats(UUID tenantId, LocalDate date) {
        Long total = reservationRepository.countByTenantIdAndFechaAndEstadoIn(
                tenantId, date, Arrays.asList("PENDIENTE", "CONFIRMADA", "EN_CURSO", "COMPLETADA"));
        Long confirmadas = reservationRepository.countByTenantIdAndFechaAndEstadoIn(
                tenantId, date, Arrays.asList("CONFIRMADA", "EN_CURSO"));

        return new ReservationStats(
                total != null ? total : 0,
                confirmadas != null ? confirmadas : 0);
    }

    public record ReservationStats(long totalReservas, long confirmadas) {
    }
}

File: frontend/pos-app/src/app/core/services/reservations.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '@env/environment';
import { AuthService } from '../auth/auth.service';

// ========== DTOs ==========

export interface RestaurantTable {
    id: string;
    numero: string;
    capacidad: number;
    ubicacion: string;
    estado: 'DISPONIBLE' | 'OCUPADA' | 'RESERVADA' | 'NO_DISPONIBLE';
    descripcion?: string;
    activo: boolean;
}

export interface Reservation {
    id: string;
    clienteNombre: string;
    clienteTelefono?: string;
    clienteEmail?: string;
    fecha: string;
    hora: string;
    personas: number;
    table?: RestaurantTable;
    estado: 'PENDIENTE' | 'CONFIRMADA' | 'EN_CURSO' | 'COMPLETADA' | 'CANCELADA' | 'NO_SHOW';
    notas?: string;
    recordatorioEnviado: boolean;
    createdAt: string;
}

export interface ReservationStats {
    totalReservas: number;
    confirmadas: number;
}

export interface CreateReservationRequest {
    clienteNombre: string;
    clienteTelefono?: string;
    fecha: string;
    hora: string;
    personas: number;
    tableId?: string;
    notas?: string;
    serviceType?: string;
}

@Injectable({
    providedIn: 'root'
})
export class ReservationsService {
    private baseUrl = environment.apiUrl || 'http://localhost:8080/api';

    constructor(
        private http: HttpClient,
        private authService: AuthService
    ) { }

    private getHeaders() {
        return {
            'X-Tenant-ID': this.authService.getTenantId() || '',
            'X-Branch-ID': this.authService.getBranchId() || '',
            'Authorization': `Bearer ${this.authService.getToken()}`
        };
    }

    // ========== RESERVATIONS ==========

    getReservations(date?: string): Observable<Reservation[]> {
        let params = new HttpParams();
        if (date) {
            params = params.set('date', date);
        }
        return this.http.get<Reservation[]>(`${this.baseUrl}/reservations`, {
            headers: this.getHeaders(),
            params
        });
    }

    getReservation(id: string): Observable<Reservation> {
        return this.http.get<Reservation>(`${this.baseUrl}/reservations/${id}`, {
            headers: this.getHeaders()
        });
    }

    createReservation(request: CreateReservationRequest): Observable<Reservation> {
        return this.http.post<Reservation>(`${this.baseUrl}/reservations`, request, {
            headers: this.getHeaders()
        });
    }

    updateStatus(id: string, estado: string): Observable<Reservation> {
        return this.http.put<Reservation>(`${this.baseUrl}/reservations/${id}/status`, { estado }, {
            headers: this.getHeaders()
        });
    }

    assignTable(id: string, tableId: string): Observable<Reservation> {
        return this.http.put<Reservation>(`${this.baseUrl}/reservations/${id}/table`, { tableId }, {
            headers: this.getHeaders()
        });
    }

    // ========== TABLES ==========

    getTables(): Observable<RestaurantTable[]> {
        return this.http.get<RestaurantTable[]>(`${this.baseUrl}/reservations/tables`, {
            headers: this.getHeaders()
        });
    }

    getAvailableTables(minCapacity = 1): Observable<RestaurantTable[]> {
        return this.http.get<RestaurantTable[]>(`${this.baseUrl}/reservations/tables/available`, {
            headers: this.getHeaders(),
            params: { minCapacity: minCapacity.toString() }
        });
    }

    updateTableStatus(id: string, estado: string): Observable<RestaurantTable> {
        return this.http.put<RestaurantTable>(`${this.baseUrl}/reservations/tables/${id}/status`, { estado }, {
            headers: this.getHeaders()
        });
    }

    // ========== STATS ==========

    getStats(date?: string): Observable<ReservationStats> {
        let params = new HttpParams();
        if (date) {
            params = params.set('date', date);
        }
        return this.http.get<ReservationStats>(`${this.baseUrl}/reservations/stats`, {
            headers: this.getHeaders(),
            params
        });
    }
    // ========== AUTOMATIONS ==========

    getAutomations(): Observable<any[]> {
        return this.http.get<any[]>(`${this.baseUrl}/automations`, {
            headers: this.getHeaders()
        });
    }

    saveAutomation(automation: any): Observable<any> {
        return this.http.post<any>(`${this.baseUrl}/automations`, automation, {
            headers: this.getHeaders()
        });
    }

    getAutomationLogs(id: string): Observable<any[]> {
        return this.http.get<any[]>(`${this.baseUrl}/automations/${id}/logs`, {
            headers: this.getHeaders()
        });
    }

    testConnection(config: any): Observable<any> {
        return this.http.post<any>(`${this.baseUrl}/automations/test`, config, {
            headers: this.getHeaders()
        });
    }
}
